### PR TITLE
12 challenge fizzbuzz program

### LINKED ISSUE #12: Challenge: FizzBuzz program
Objective:
Create a program that prints the numbers from 1 to 100. But for multiples of three, print "Fizz" instead of the number, and for multiples of five, print "Buzz". For numbers that are multiples of both three and five, print "FizzBuzz".

Another way of putting it: 
# **FizzBuzz Challenge** 
This is a FizzBuzz program that prints numbers from 1 to 100 with specific rules: 
- For multiples of three, print "Fizz". 
- For multiples of five, print "Buzz". 
- For multiples of both three and five, print "FizzBuzz".
-->

For each number from 1 to 100:
    If the number is divisible by both 3 and 5:
        Print "FizzBuzz"
    Else if the number is divisible by 3:
        Print "Fizz"
    Else if the number is divisible by 5:
        Print "Buzz"
    Else:
        Print the number

Resource Guide for Solving "FizzBuzz" Challenge
Understanding Loops and Conditionals:

Python For Loops - W3Schools  https://www.w3schools.com/python/python_for_loops.asp?form=MG0AV3

Python If...Else Statements - W3Schools   https://www.w3schools.com/python/python_conditions.asp?form=MG0AV3

Modulo Operator:

Python Operators - W3Schools  https://www.w3schools.com/python/python_operators.asp?form=MG0AV3

Focus on the "Arithmetic Operators" section, especially the modulo operator (%).

These articles provide a general approach to breaking down and solving the problem.

### PR DESCRIPTION
**Pull request by hectordacb**
## Behavior
Objective:
Create a program that prints the numbers from 1 to 100. But for multiples of three, print "Fizz" instead of the number, and for multiples of five, print "Buzz". For numbers that are multiples of both three and five, print "FizzBuzz".

Another way of putting it: 
# **FizzBuzz Challenge** 
This is a FizzBuzz program that prints numbers from 1 to 100 with specific rules: 
- For multiples of three, print "Fizz". 
- For multiples of five, print "Buzz". 
- For multiples of both three and five, print "FizzBuzz".
-->

For each number from 1 to 100:
    If the number is divisible by both 3 and 5:
        Print "FizzBuzz"
    Else if the number is divisible by 3:
        Print "Fizz"
    Else if the number is divisible by 5:
        Print "Buzz"
    Else:
        Print the number

Resource Guide for Solving "FizzBuzz" Challenge
Understanding Loops and Conditionals:

Python For Loops - W3Schools  https://www.w3schools.com/python/python_for_loops.asp?form=MG0AV3

Python If...Else Statements - W3Schools   https://www.w3schools.com/python/python_conditions.asp?form=MG0AV3

Modulo Operator:

Python Operators - W3Schools  https://www.w3schools.com/python/python_operators.asp?form=MG0AV3

Focus on the "Arithmetic Operators" section, especially the modulo operator (%).

These articles provide a general approach to breaking down and solving the problem.
### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/challenge_12/test_fizzbuzz_program.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import unittest
+
+from solutions.challenge_12.fizzbuzz_program import fizzbuzz
+
+
+class TestFizzBuzz(unittest.TestCase):
+    """Unit tests for the fizzbuzz function."""
+
+    def test_fizzbuzz_case_first(self):
```
Comment (RamonColmenares) [inline solutions/tests/challenge_12/test_fizzbuzz_program.py]: The test has to have a helpful name, describing what are you testing.

this has to be applied to all the tests name

File: solutions/challenge_12/fizzbuzz_program.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        >>> fizzbuzz(3)
+        ['1', '2', 'Fizz']
+
+        >>> fizzbuzz(5)
+        ['1', '2', 'Fizz', '4', 'Buzz']
+
+        >>> fizzbuzz(15)[-1]
+        'FizzBuzz'
+    """
+    if not isinstance(n, int) or n <= 0:
```
Comment (RamonColmenares) [inline solutions/challenge_12/fizzbuzz_program.py]: The defensive assertions are documented using Raises:
 Each assumption about an argument is checked with an assertion
 Each assertion checks for only one assumption about the argument
 
 Please check this on the assertion

**Review by RamonColmenares — CHANGES_REQUESTED** [review]
Hi Hector, please check the review comments. Thanks

File: solutions/tests/challenge_12/test_fizzbuzz_program.py (hunk 1 above)
Comment (hectordacb) [reply solutions/tests/challenge_12/test_fizzbuzz_program.py]: This was fixed, now the tests have more helpful names

File: solutions/challenge_12/fizzbuzz_program.py (hunk 2 above)
Comment (hectordacb) [reply solutions/challenge_12/fizzbuzz_program.py]: The assertion for the data type was changed to use assert

**Review by RamonColmenares — APPROVED** [review]
LGTM!
